metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/sql/DirectSqlTable.java

```diff
@@ -169,56 +176,93 @@ public boolean exists(final QualifiedName name) {
+    public void updateIcebergTable(final TableInfo tableInfo) {
+        final QualifiedName tableName = tableInfo.getName();
+        final Long tableId = getTableId(tableName);
+        Map<String, String> existingTableMetadata = Maps.newHashMap();
+        log.debug("Lock Iceberg table {}", tableName);
         try {
-            tableType = jdbcTemplate.queryForObject(SQL.TABLE_PARAM_LOCK,
-                new SqlParameterValue[] {new SqlParameterValue(Types.BIGINT, tableId),
-                    new SqlParameterValue(Types.VARCHAR, PARAM_TABLE_TYPE), }, String.class);
-        } catch (EmptyResultDataAccessException ignored) { }
-        if (tableType == null || !ICEBERG_TABLE_TYPE.equalsIgnoreCase(tableType)) {
-            final String message = String.format("Originally table %s is not of type iceberg", tableName);
-            log.info(message);
+            existingTableMetadata = jdbcTemplate.query(SQL.TABLE_PARAMS_LOCK,
+                new SqlParameterValue[]{new SqlParameterValue(Types.BIGINT, tableId)}, rs -> {
+                    final Map<String, String> result = Maps.newHashMap();
+                    while (rs.next()) {
+                        result.put(rs.getString(COL_PARAM_KEY), rs.getString(COL_PARAM_VALUE));
+                    }
+                    return result;
+                });
+        } catch (EmptyResultDataAccessException ex) {
+            log.info(String.format("No parameters defined for iceberg table %s", tableName));
+        } catch (Exception ex) {
+            final String message = String.format("Failed getting a lock on iceberg table %s", tableName);
+            log.warn(message, ex);
             throw new InvalidMetaException(tableName, message, null);
         }
-        Boolean isLocked = null;
-        try {
-            isLocked = jdbcTemplate.queryForObject(SQL.TABLE_PARAM_LOCK,
-                new SqlParameterValue[] {new SqlParameterValue(Types.BIGINT, tableId),
-                    new SqlParameterValue(Types.VARCHAR, PARAM_METADATA_LOCK), }, Boolean.class);
-        } catch (EmptyResultDataAccessException ignored) { }
-        if (isLocked == null) {
-            insertTableParam(tableId, PARAM_METADATA_LOCK, 1);
-        } else if (isLocked) {
-            final String message = String.format("Iceberg table %s is locked.", tableName);
+        if (existingTableMetadata.isEmpty()
+            || !ICEBERG_TABLE_TYPE.equalsIgnoreCase(existingTableMetadata.get(PARAM_TABLE_TYPE))) {
+            final String message = String.format("Originally table %s is not of type iceberg", tableName);
             log.info(message);
-            throw new IllegalStateException(message);
-        } else {
-            updateTableParam(tableId, PARAM_METADATA_LOCK, 1);
+            throw new InvalidMetaException(tableName, message, null);
         }
+        final Map<String, String> newTableMetadata = tableInfo.getMetadata() == null
+            ? Maps.newHashMap() : tableInfo.getMetadata();
```

Comment:
When will tableInfo.getMetadata() be null? This function is invoked for iceberg table only, which is checked from HiveTableUtil.isIcebergTable(tableInfo). If tableInfo.getMetadata() is null here, something is wrong. We can put a check to ensure not null case or at the very beginning of this function if having the null concern.

Reply:
Good point. Since this is a public method is a separate class, I added it as a safety.

Reply:
We should fail in this case, otherwise the following validateIcebergUpdate() function will fail with a miss-leading exception message.

Reply:
It will fail saying the metadata location is empty. Are you suggesting, fail even before locking the records?